File: windows/extension/youtube-normalize.js
/**
 * YouTube canonical-object normalization — JS port for the Windows MV3 extension.
 *
 * ⚠️ LOCKSTEP: This is a faithful, dependency-free port of
 * `shared/youtube/youtube-normalize.ts`. The TypeScript file is the SPEC and the
 * source of truth; this port MUST produce identical canonical objects and policy
 * keys for the same input. Any change to the shared file must be mirrored here
 * (enforced by review). Do not "improve" one side independently.
 *
 * The product treats YouTube resources as FIRST-CLASS POLICY TARGETS, not URL
 * strings. Every enforcement adapter (iOS NEFilterDataProvider, macOS Safari Web
 * Extension, this Windows extension, the NEURLFilter dataset builder) reduces an
 * observed URL to the same canonical object BEFORE consulting policy, so a parent
 * who approves `YOUTUBE_VIDEO:dQw4w9WgXcQ` approves the video no matter which URL
 * form the child's browser used.
 *
 * Approving a single video must NOT implicitly approve the channel,
 * recommendations, related videos, search, comments, Shorts, or arbitrary
 * navigation. That guarantee lives in the POLICY ENGINE (evaluation order), not
 * here; this module only answers "what canonical object does this URL denote?".
 */

/**
 * @typedef {("video"|"channel"|"playlist"|"watch_with_playlist"|"search"|"shorts"|"other")} YouTubeObjectKind
 */

/**
 * @typedef {Object} YouTubeObject
 * @property {YouTubeObjectKind} kind
 * @property {string} [videoId]      Canonical 11-char video id.
 * @property {string} [channelId]    Canonical channel id (UC…).
 * @property {string} [channelHandle] @handle or legacy user//c/ token.
 * @property {string} [playlistId]   Canonical playlist id (PL…, UU…, LL…, FL…, RD…).
 * @property {boolean} isYouTube     True when the host/path is a YouTube surface at all.
 * @property {string} [host]         The host we matched, normalized (no leading www.).
 */

/** Hosts that serve the YouTube application surface (watch pages, shorts, etc.). */
const YT_APP_HOSTS = new Set([
  "youtube.com",
  "m.youtube.com",
  "music.youtube.com",
  "gaming.youtube.com",
  "youtubekids.com",
  "youtube-nocookie.com", // privacy-enhanced embeds
]);

/** Short-link host: the entire path is the video id. */
const YT_SHORT_HOSTS = new Set(["youtu.be"]);

/**
 * Resource hosts REQUIRED for an approved video to actually play. These are NOT
 * policy targets — they are documented here so the extension never blocks them
 * when a video is approved. A default-deny YouTube policy that also blocks these
 * will show an approved video that spins forever. See ARCHITECTURE.md §6.
 *
 * NOTE: googlevideo.com is the media CDN; its URLs are opaque, per-session, and
 * cannot be tied to a specific video id from the URL alone, so the adapter must
 * allow the whole host when ANY video is currently approved on the device, and
 * rely on the watch-page gate (which IS per-video) to control access. Deliberate,
 * documented limitation — kept identical to the shared TS.
 */
export const YOUTUBE_PLAYBACK_SUPPORT_HOSTS = [
  "www.youtube.com", // InnerTube /youtubei/v1/player API, base JS
  "youtubei.googleapis.com", // InnerTube API (mobile/native players)
  "i.ytimg.com", // thumbnails / storyboards
  "s.ytimg.com", // player javascript/css
  "yt3.ggpht.com", // avatars/thumbs
  "*.googlevideo.com", // MEDIA STREAMS (opaque; see note above)
  "jnn-pa.googleapis.com", // BotGuard/attestation used by the player
  "fonts.gstatic.com", // player glyphs
];

const VIDEO_ID_RE = /^[A-Za-z0-9_-]{11}$/;
const CHANNEL_ID_RE = /^UC[A-Za-z0-9_-]{22}$/;
const PLAYLIST_ID_RE = /^(?:PL|UU|LL|FL|RD|OL|EL)[A-Za-z0-9_-]{10,}$/;

function stripWww(host) {
  return host.replace(/^www\./i, "").toLowerCase();
}

function isValidVideoId(id) {
  return !!id && VIDEO_ID_RE.test(id);
}

/**
 * Reduce a URL to its canonical YouTube object. Returns `{ isYouTube:false }`
 * for non-YouTube URLs. Never throws.
 * @param {string} rawUrl
 * @returns {YouTubeObject}
 */
export function normalizeYouTube(rawUrl) {
  let u;
  try {
    u = new URL(rawUrl);
  } catch {
    return { kind: "other", isYouTube: false };
  }

  const host = stripWww(u.hostname);
  const notYt = { kind: "other", isYouTube: false, host };

  // youtu.be/<id>
  if (YT_SHORT_HOSTS.has(host)) {
    const seg = u.pathname.split("/").filter(Boolean);
    const id = seg[0];
    if (isValidVideoId(id)) {
      const listId = u.searchParams.get("list") ?? undefined;
      return {
        kind: listId ? "watch_with_playlist" : "video",
        videoId: id,
        playlistId: listId && PLAYLIST_ID_RE.test(listId) ? listId : undefined,
        isYouTube: true,
        host,
      };
    }
    return { ...notYt, isYouTube: true };
  }

  if (!YT_APP_HOSTS.has(host)) return notYt;

  const path = u.pathname.replace(/\/+$/, ""); // trim trailing slashes
  const seg = path.split("/").filter(Boolean);
  const p0 = seg[0]?.toLowerCase();

  // /watch?v=<id>[&list=<pl>]
  if (p0 === "watch") {
    const id = u.searchParams.get("v");
    const listId = u.searchParams.get("list") ?? undefined;
    if (isValidVideoId(id)) {
      return {
        kind: listId ? "watch_with_playlist" : "video",
        videoId: id,
        playlistId: listId && PLAYLIST_ID_RE.test(listId) ? listId : undefined,
        isYouTube: true,
        host,
      };
    }
    return { ...notYt, isYouTube: true };
  }

  // /shorts/<id>
  if (p0 === "shorts" && isValidVideoId(seg[1])) {
    return { kind: "shorts", videoId: seg[1], isYouTube: true, host };
  }

  // /embed/<id>  and /v/<id>  (nocookie host lands here too)
  if ((p0 === "embed" || p0 === "v") && isValidVideoId(seg[1])) {
    return { kind: "video", videoId: seg[1], isYouTube: true, host };
  }
  // /embed/videoseries?list=<pl>
  if (p0 === "embed" && seg[1]?.toLowerCase() === "videoseries") {
    const listId = u.searchParams.get("list") ?? undefined;
    if (listId && PLAYLIST_ID_RE.test(listId))
      return { kind: "playlist", playlistId: listId, isYouTube: true, host };
  }

  // /live/<id>  (premieres / livestream permalinks)
  if (p0 === "live" && isValidVideoId(seg[1])) {
    return { kind: "video", videoId: seg[1], isYouTube: true, host };
  }

  // /playlist?list=<pl>
  if (p0 === "playlist") {
    const listId = u.searchParams.get("list") ?? undefined;
    if (listId && PLAYLIST_ID_RE.test(listId))
      return { kind: "playlist", playlistId: listId, isYouTube: true, host };
    return { ...notYt, isYouTube: true };
  }

  // /channel/<UC…>
  if (p0 === "channel" && seg[1] && CHANNEL_ID_RE.test(seg[1])) {
    return { kind: "channel", channelId: seg[1], isYouTube: true, host };
  }

  // /@handle  (modern channel handle)
  if (p0 && p0.startsWith("@")) {
    return { kind: "channel", channelHandle: seg[0], isYouTube: true, host };
  }

  // /user/<name>, /c/<name>  (legacy channel forms — resolve to channelId
  // server-side via the Data API; carried as a handle-like token meanwhile)
  if ((p0 === "user" || p0 === "c") && seg[1]) {
    return { kind: "channel", channelHandle: `${p0}/${seg[1]}`, isYouTube: true, host };
  }

  // /results?search_query=…
  if (p0 === "results") {
    return { kind: "search", isYouTube: true, host };
  }

  // Recognized YouTube surface we can't reduce to a specific object
  return { kind: "other", isYouTube: true, host };
}

/**
 * Stable policy key for a canonical object, e.g. "YOUTUBE_VIDEO:dQw4w9WgXcQ".
 * Kept identical to `youTubePolicyKey` in the shared TS.
 * @param {YouTubeObject} o
 * @returns {string|null}
 */
export function youTubePolicyKey(o) {
  if (o.videoId) return `YOUTUBE_VIDEO:${o.videoId}`;
  if (o.playlistId && o.kind === "playlist") return `YOUTUBE_PLAYLIST:${o.playlistId}`;
  if (o.channelId) return `YOUTUBE_CHANNEL:${o.channelId}`;
  if (o.channelHandle) return `YOUTUBE_CHANNEL_HANDLE:${o.channelHandle}`;
  return null;
}

/**
 * Returns true if `host` is (or is a subdomain of, for the wildcard entry) one of
 * the playback-support hosts. Used so the extension never blocks the resources an
 * approved video needs to stream.
 * @param {string} host
 * @returns {boolean}
 */
export function isPlaybackSupportHost(host) {
  const h = stripWww(host);
  for (const entry of YOUTUBE_PLAYBACK_SUPPORT_HOSTS) {
    if (entry.startsWith("*.")) {
      const base = entry.slice(2);
      if (h === base || h.endsWith(`.${base}`)) return true;
    } else if (h === stripWww(entry)) {
      return true;
    }
  }
  return false;
}

File: apple/AjarFilter/Shared/YouTubeNormalize.swift
import Foundation

/// Swift mirror of `shared/youtube/youtube-normalize.ts`. The TypeScript file is
/// the authoritative spec; this must produce identical canonical objects so the
/// iOS content filter keys policy the same way as the backend and other adapters.
///
/// Approving one video must never widen to channel/recommendations/Shorts/search
/// — that guarantee is in the policy evaluation order, not here. This only maps a
/// URL to the canonical object it denotes.

public enum YouTubeKind: String {
    case video, channel, playlist, watchWithPlaylist, search, shorts, other
}

public struct YouTubeObject {
    public var kind: YouTubeKind
    public var videoId: String?
    public var channelId: String?
    public var channelHandle: String?
    public var playlistId: String?
    public var isYouTube: Bool
    public var host: String?
}

public enum YouTube {

    /// Hosts serving the YouTube application surface.
    static let appHosts: Set<String> = [
        "youtube.com", "m.youtube.com", "music.youtube.com",
        "gaming.youtube.com", "youtubekids.com", "youtube-nocookie.com",
    ]
    static let shortHosts: Set<String> = ["youtu.be"]

    /// Hosts required for an APPROVED video to actually play. Never block these
    /// while a video is approved. `*.googlevideo.com` is the opaque media CDN;
    /// allow the host while any video is approved and gate access at the watch
    /// page (which is per-video). See ARCHITECTURE.md §YouTube.
    public static let playbackSupportHosts: [String] = [
        "www.youtube.com", "youtubei.googleapis.com", "i.ytimg.com",
        "s.ytimg.com", "yt3.ggpht.com", "*.googlevideo.com",
        "jnn-pa.googleapis.com", "fonts.gstatic.com",
    ]

    /// Paths on `www.youtube.com` that are genuinely player plumbing.
    ///
    /// That host serves the InnerTube API *and* every watch page, search page
    /// and Short. Treating the whole host as playback support would mean that
    /// while ANY video is approved, every other video and all of YouTube search
    /// is allowed — the product's headline guarantee, silently void. The Windows
    /// extension carries this same list and the same warning.
    static let playerPathPrefixes: [String] = [
        "/youtubei/",       // InnerTube player API
        "/s/player/",       // base player JS/CSS
        "/api/timedtext",   // captions
        "/videoplayback",   // media (also served from googlevideo)
        "/generate_204", "/error_204",  // beacons
    ]

    /// Does this host serve playback plumbing? Handles the one wildcard entry.
    static func isPlaybackSupportHost(_ rawHost: String) -> Bool {
        let host = Host.normalize(rawHost)
        for entry in playbackSupportHosts {
            if entry.hasPrefix("*.") {
                let suffix = String(entry.dropFirst(2))
                if host == suffix || host.hasSuffix(".\(suffix)") { return true }
            } else if host == Host.normalize(entry) {
                return true
            }
        }
        return false
    }

    /// Is this request the PLUMBING an already-approved video needs, rather than
    /// a page?
    ///
    /// `pathIsKnown` is false for a socket flow, where the provider has a
    /// hostname and nothing else. On the page host that is decisive: without a
    /// path we cannot tell `/youtubei/v1/player` from `/watch?v=…`, so the host
    /// must NOT qualify — otherwise one approved video opens all of YouTube
    /// through the very carve-out meant to keep it shut. The other hosts serve
    /// only plumbing, so a hostname alone is enough for them.
    public static func isPlaybackSupport(host: String, path: String?, pathIsKnown: Bool) -> Bool {
        guard isPlaybackSupportHost(host) else { return false }
        let bare = Host.normalize(host)
        let isPageHost = bare == "youtube.com" || bare == "www.youtube.com"
        guard isPageHost else { return true }
        guard pathIsKnown, let path else { return false }
        return playerPathPrefixes.contains { path.hasPrefix($0) }
    }

    static func isVideoId(_ s: String?) -> Bool {
        guard let s else { return false }
        return s.range(of: "^[A-Za-z0-9_-]{11}$", options: .regularExpression) != nil
    }
    static func isChannelId(_ s: String) -> Bool {
        s.range(of: "^UC[A-Za-z0-9_-]{22}$", options: .regularExpression) != nil
    }
    static func isPlaylistId(_ s: String) -> Bool {
        s.range(of: "^(?:PL|UU|LL|FL|RD|OL|EL)[A-Za-z0-9_-]{10,}$", options: .regularExpression) != nil
    }
    /// The TS `stripWww` is `host.replace(/\.$/, "").replace(/^www\./i, "").toLowerCase()`
    /// — it strips the trailing ROOT DOT as well. The previous Swift version did
    /// not, so `https://youtube.com./watch?v=…` was not recognized as YouTube at
    /// all and fell through to the web default. Delegates to `Host.normalize`,
    /// which is that exact expression.
    static func stripWww(_ host: String) -> String { Host.normalize(host) }

    public static func normalize(_ raw: String) -> YouTubeObject {
        guard let u = URLComponents(string: raw), let hostRaw = u.host else {
            return YouTubeObject(kind: .other, videoId: nil, channelId: nil,
                                 channelHandle: nil, playlistId: nil, isYouTube: false, host: nil)
        }
        let host = stripWww(hostRaw)
        func q(_ name: String) -> String? { u.queryItems?.first(where: { $0.name == name })?.value }
        let notYt = YouTubeObject(kind: .other, videoId: nil, channelId: nil,
                                  channelHandle: nil, playlistId: nil, isYouTube: false, host: host)

        // youtu.be/<id>
        if shortHosts.contains(host) {
            let seg = u.path.split(separator: "/").map(String.init)
            if let id = seg.first, isVideoId(id) {
                let list = q("list")
                return YouTubeObject(kind: list != nil ? .watchWithPlaylist : .video,
                                     videoId: id, channelId: nil, channelHandle: nil,
                                     playlistId: (list.map(isPlaylistId) == true) ? list : nil,
                                     isYouTube: true, host: host)
            }
            var o = notYt; o.isYouTube = true; return o
        }

        guard appHosts.contains(host) else { return notYt }

        let path = u.path.hasSuffix("/") ? String(u.path.dropLast()) : u.path
        let seg = path.split(separator: "/").map(String.init)
        let p0 = seg.first?.lowercased()

        switch p0 {
        case "watch":
            if let id = q("v"), isVideoId(id) {
                let list = q("list")
                return YouTubeObject(kind: list != nil ? .watchWithPlaylist : .video,
                                     videoId: id, channelId: nil, channelHandle: nil,
                                     playlistId: (list.map(isPlaylistId) == true) ? list : nil,
                                     isYouTube: true, host: host)
            }
        case "shorts" where isVideoId(seg[safe: 1]):
            return YouTubeObject(kind: .shorts, videoId: seg[1], channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case "embed", "v":
            if isVideoId(seg[safe: 1]) {
                return YouTubeObject(kind: .video, videoId: seg[1], channelId: nil, channelHandle: nil,
                                     playlistId: nil, isYouTube: true, host: host)
            }
            if p0 == "embed", seg[safe: 1]?.lowercased() == "videoseries", let list = q("list"), isPlaylistId(list) {
                return YouTubeObject(kind: .playlist, videoId: nil, channelId: nil, channelHandle: nil,
                                     playlistId: list, isYouTube: true, host: host)
            }
        case "live" where isVideoId(seg[safe: 1]):
            return YouTubeObject(kind: .video, videoId: seg[1], channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case "playlist":
            if let list = q("list"), isPlaylistId(list) {
                return YouTubeObject(kind: .playlist, videoId: nil, channelId: nil, channelHandle: nil,
                                     playlistId: list, isYouTube: true, host: host)
            }
        case "channel" where seg[safe: 1].map(isChannelId) == true:
            return YouTubeObject(kind: .channel, videoId: nil, channelId: seg[1], channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case let p? where p.hasPrefix("@"):
            return YouTubeObject(kind: .channel, videoId: nil, channelId: nil, channelHandle: seg[0],
                                 playlistId: nil, isYouTube: true, host: host)
        case "user", "c":
            if let name = seg[safe: 1] {
                return YouTubeObject(kind: .channel, videoId: nil, channelId: nil,
                                     channelHandle: "\(p0!)/\(name)", playlistId: nil, isYouTube: true, host: host)
            }
        case "results":
            return YouTubeObject(kind: .search, videoId: nil, channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        default:
            break
        }
        var o = notYt; o.isYouTube = true; return o
    }

    /// Stable policy key, e.g. "YOUTUBE_VIDEO:dQw4w9WgXcQ".
    public static func policyKey(_ o: YouTubeObject) -> String? {
        if let v = o.videoId { return "YOUTUBE_VIDEO:\(v)" }
        if let pl = o.playlistId, o.kind == .playlist { return "YOUTUBE_PLAYLIST:\(pl)" }
        if let c = o.channelId { return "YOUTUBE_CHANNEL:\(c)" }
        if let h = o.channelHandle { return "YOUTUBE_CHANNEL_HANDLE:\(h)" }
        return nil
    }
}

private extension Array {
    subscript(safe i: Int) -> Element? { indices.contains(i) ? self[i] : nil }
}
